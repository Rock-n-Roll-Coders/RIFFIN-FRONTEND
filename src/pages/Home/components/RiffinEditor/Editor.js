// Components / hooks
import AddNewBlockButton from "./components/AddNewBlockButton/AddNewBlockButton";
import TitleInput from "./components/TitleInput/TitleInput";
import DeleteTabButton from "./components/DeleteTabButton/DeleteTabButton";
import SaveTabButton from "./components/SaveTabButton/SaveTabButton";
import TablatureBlock from "./components/TablatureBlock/TablatureBlock";
import { useContext } from "react";
import { RiffinEditorDispatch } from "./RiffinProvider";
// MUI
import { Box, Grid } from "@mui/material";
const Content = () => {
  const { editor } = useContext(RiffinEditorDispatch);
  return (
    <>
        <Grid container rowSpacing={2} columnSpacing={4} sx={{alignItems: "end"}}>
          <Grid item>
            <TitleInput />
          </Grid>
          <Grid item>
            <SaveTabButton />
          </Grid>
          <Grid item>
            <DeleteTabButton />
          </Grid>
        </Grid>
        {editor.tablature.blocks.map((block, i) => (
          <Box sx={{my: 4}} key={i}>
            <TablatureBlock key={i} index={i} block={block} />
          </Box>
        ))}
        <AddNewBlockButton />
    </>
  );
}
 
export default Content;