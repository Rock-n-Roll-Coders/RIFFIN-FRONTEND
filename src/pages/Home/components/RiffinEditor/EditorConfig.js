/**
 * How many columns are between steps in the SizeSlider.
 */
export const STEP_COUNT = 5;

/**
 * Rows of the textarea when editing a bass tab.
 */
export const BASS_STRING_COUNT = 4;

/**
 * Rows of the textarea when editing a guitar tab.
 */
export const GUITAR_STRING_COUNT = 6;

/**
 * Minimum allowable columns the textarea can be sized down to.
 */
export const MIN_BLOCK_COLS = 20;

/**
 * Maximum allowable columns the textarea can be sized
 */
export const MAX_BLOCK_COLS = 80;

/**
 * The default amount of columns that a block starts with.
 */
export const DEFAULT_BLOCK_COLS = 40;

/**
 * Controls how many columns to skip when duplicating a column. A target of 2 looks as follows:
 * [x][ ][ ] (duplicate) => [x][ ][x]
 */
export const DEFAULT_DUPLICATION_TARGET = 2;

/**
 * Map of allowable movement keys. The value is not used.
 */
export const MOVEMENT_KEYS = {
  ArrowDown: true,
  ArrowLeft: true,
  ArrowRight: true,
  ArrowUp: true,
};

/**
 * Key/Value pairs of allowable inputs. The "key" is the pressed keyboard key and the "value" is the dispatch action associated with the key.
 */
export const LEGAL_INPUTS = {
  "~": "addCharacter", // vibrato
  "/": "addCharacter", // slide
  "^": "addCharacter", // bend
  "x": "addCharacter", // mute
  "p": "addCharacter", // pull off
  "h": "addCharacter", // hammer on
  0: "addCharacter",
  1: "addCharacter",
  2: "addCharacter",
  3: "addCharacter",
  4: "addCharacter",
  5: "addCharacter",
  6: "addCharacter",
  7: "addCharacter",
  8: "addCharacter",
  9: "addCharacter",
  "]": "duplicateColumn", // duplicate chord
  "[": "deleteColumn", // deleteChord
  "Backspace": "deleteCharacter",
};

/**
 * Maximum characters in the note that sits above each TablatureBlock
 */
export const MAX_NOTE_CHARACTERS = 256;

/**
 * Maximum number of TablatureBlocks a user can have per document.
 */
export const MAX_BLOCKS = 12;