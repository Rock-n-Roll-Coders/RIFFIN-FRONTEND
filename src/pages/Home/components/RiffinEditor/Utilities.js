import { DEFAULT_BLOCK_COLS, MIN_BLOCK_COLS, MAX_BLOCK_COLS, MOVEMENT_KEYS, LEGAL_INPUTS } from './EditorConfig'

const calcNewMaxLength = (numberOfCols, numberOfStrings) => (numberOfCols * numberOfStrings) + (numberOfStrings * 2) - 1
const calculateNewFirstCol = (numberOfCols, index) => (numberOfCols + 1) * (index)
const calculateNewLastCol = (numberOfCols, stringNum, index) => (numberOfCols * stringNum) + index

/* 
  Needs an actions object with the following properties:
  area
  characterToAdd
  action
  cols,
  stepCount
*/
export const updateBlockValue = (action) => {
  if((action.type === "increase" && action.cols === MAX_BLOCK_COLS) ||
    (action.type === "decrease" && action.cols === MIN_BLOCK_COLS)) {
    return action.area
  }

  const guitarStrings = action.area.split('\n')
  let newValueString = ""
  for(let i = 0; i < guitarStrings.length; i++) {
    const array = [...guitarStrings[i]]
    for(let j = 0; j < Math.abs(action.stepCount); j++) {
      if(action.type === "increase") {
        array[array.length - 1] = action.characterToAdd
        array.push('|')
      } else {
        array.pop()
        array[array.length - 1] = "|"
      }
    }
    if(i !== guitarStrings.length - 1) { // Last string doesn't need a newline char
      array.push("\n")
    }
    newValueString += array.join("")
  }
  return newValueString
}

/* 
  action {
    cols,
    maxLength,
    action,
    stringCount,
    stepCount
  }
*/
export const updateTextAreaAttributes = (action) => {
  if((action.type === "increase" && action.cols === MAX_BLOCK_COLS) ||
    (action.type === "decrease" && action.cols === MIN_BLOCK_COLS)) {
    return {
      cols: action.cols,
      maxLength: action.maxLength
    }
  }

  return {
    cols: action.cols + action.stepCount,
    maxLength: calcNewMaxLength(action.cols + action.stepCount, action.stringCount)
  }

}

export const getMapOfLastColumnIndexes = (action) => {
  const newLastCols = {}
  for(let i = 0; i < action.stringCount; i++) {
    const stringNum = i + 1
    const newVal = calculateNewLastCol(action.cols, stringNum, i)
    newLastCols[newVal] = true
  }
  return newLastCols
}

export const getMapOfFirstColumnIndexes = (action) => {
  const newFirstCols = {}
  for(let i = 0; i < action.stringCount; i++) {
    const newVal = calculateNewFirstCol(action.cols, i)
    newFirstCols[newVal] = true
  }
  return newFirstCols
}

// TODO Rename to createNewBlock
export const getNewGuitarBlock = (stringCount = 6) => {
  const action = {
    stringCount: stringCount,
    cols: DEFAULT_BLOCK_COLS
  }
  const mapOfFirstColumnIndexes = getMapOfFirstColumnIndexes(action)
  const mapOfLastColumnIndexes = getMapOfLastColumnIndexes(action)
  const mapOfSecondToLastColumnIndexes = Object.keys(mapOfLastColumnIndexes).map((val) => val - 1)
  const initLength = (action.stringCount * action.cols) + (action.stringCount - 1)
  const initTextAreaWithValue = (character) => {
    let charactersInString = [];
    for (let i = 0; i < initLength; i++) {
      if (i in mapOfLastColumnIndexes) {
        charactersInString.push("\n");
      } else if (i in mapOfFirstColumnIndexes || mapOfSecondToLastColumnIndexes.includes(i)) {
        charactersInString.push("|");
      } else {
        charactersInString.push(character);
      }
    }
    return charactersInString.join("");
  };
  const inputs = initTextAreaWithValue(" ")
  const dashes = initTextAreaWithValue("-")
  const maxLength = calcNewMaxLength(action.cols, action.stringCount)
  return {
    tempKey: Date() + Math.random(),
    inputs: inputs,
    dashes: dashes,
    cols: DEFAULT_BLOCK_COLS,
    maxLength: maxLength,
    numberOfStrings: stringCount
  };
}

// ----------------------------------------------------------------

/**
 * Creates the initial tablature object for a new tab
 * @param {number} numberOfStrings 
 * @returns a tablature object
 */
export const getNewTablatureTemplateObject = (numberOfStrings) => {
  const initialBlock = getNewGuitarBlock(numberOfStrings);
  return {
    name: "A tasty lick",
    blocks: [initialBlock],
    tags: [],
    numberOfStrings
  }
}

/**
 * Mainly used for compositional purposes. Creates a selection object with the provided arguments.
 * @param {Number} blockIndex 
 * @param {Ref} blockRef 
 * @returns the selection object
 */
export const generateSelectedBlockObject = (blockIndex, blockRef) => ({ inputRef: blockRef, index: blockIndex })

/**
 * Mainly used for compositional purposes. Creates a cursor object with the provided arguments.
 * @param {Number} selectionStart 
 * @returns the cursorPosition object
 */
export const generateCursorPositionObject = (selectionStart) => ({ position: selectionStart })

/**
 * Checks if the provided key controls movement
 * @param {String} key 
 * @returns true or false depending on if the key is a movement key
 */
export const isMovementKey = (key) => (key in MOVEMENT_KEYS)

/**
 * Checks if the key is a legal input (other than arrow keys) and returns the dispatch type associated with that key.
 * @param {String} key 
 * @returns The dispatch type associated with the key, or undegined if the key is not a legal input.
 */
export const getDispatchTypeOfPressedKey = (key) => (key in LEGAL_INPUTS) ? LEGAL_INPUTS[key] : undefined

/**
 * Replaces a character in the provided string at the specified position.
 * @param {String} originalInputs
 * @param {String} character 
 * @param {Number} position 
 * @returns a new string with the replacement character inserted
 */
export const replaceTextareaValue = (textAreaValue, character, position) => {
  let arrayOfCharacters = [...textAreaValue];
  arrayOfCharacters[position] = character;
  return arrayOfCharacters.join("");
}

const getPositionsAboveSelection = (selectionStart, cols) => {
  let arrayOfPositions = []
  const indexGapBetweenStrings = cols + 1 // The + 1 compensates for the hidden \n
  for(let position = selectionStart - indexGapBetweenStrings; position >= 0; position -= indexGapBetweenStrings) {
    arrayOfPositions.push(position - 1)
  }
  return arrayOfPositions
}

const getPositionsBelowSelection = (selectionStart, cols, editableLength) => {
  const indexGapBetweenStrings = cols + 1 // The + 1 compensates for the hidden \n
  let arrayOfPositions = []
  for(let position = selectionStart + indexGapBetweenStrings; position <= editableLength; position += indexGapBetweenStrings) {
    arrayOfPositions.push(position - 1)
  }
  return arrayOfPositions
}

export const getPositionsToDuplicate = (selectionStart, cols, maxLength, numberOfStrings) => {
  let numberOfEditablePositionsOnBlock = maxLength - numberOfStrings
  let positionsAboveSelection = getPositionsAboveSelection(selectionStart - 1, cols)
  let positionsBelowSelection = getPositionsBelowSelection(selectionStart - 1, cols, numberOfEditablePositionsOnBlock)
  let allPositionsToDuplicate = [selectionStart - 2, ...positionsAboveSelection, ...positionsBelowSelection]
  return allPositionsToDuplicate
}

// function handleDuplicateChord(mapOfFirstColumnIndexes, mapOfLastColumnIndexes) {

//   const selectedBlock = tablature.blocks[selectedTablatureBlock.index]
//   // 4) get positions to duplicate
//   let positionsToDuplicate = getPositionsToDuplicate(cursorPosition.position, selectedBlock.cols, (selectedBlock.maxLength - tablature.numberOfStrings))

//   // 5) check if a duplication occurs
//   if(positionsToDuplicate.length === 0) {
//     return
//   }

//   // 6) replace characters
//   let inputsAsArray = [...selectedBlock.inputs]
//   positionsToDuplicate.forEach(pos => {
//     const characterToDuplicate = inputsAsArray[pos]
//     selectedBlock.inputs = getUpdatedTextAreaValues(
//       "inputs",
//       characterToDuplicate,
//       pos + 2
//     )
//     selectedBlock.dashes = getUpdatedTextAreaValues(
//       "dashes",
//       (characterToDuplicate !== " ") ? " " : "-",
//       pos + 2
//     )
//   })

//   // 7) set cursor and tablature
//   setCursorPosition({ position: cursorPosition.position + 2 })
//   refreshTablatureObject()
// }