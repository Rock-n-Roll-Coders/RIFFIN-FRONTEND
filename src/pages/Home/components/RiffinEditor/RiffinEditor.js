import { createContext, useReducer, useEffect, useState } from "react";
import * as utils from "./Utilities";
import TablatureBlock from "./components/TablatureBlock/TablatureBlock";
import AddTablatureBlockButton from "./components/AddTablatureBlockButton/AddTablatureBlockButton";
import TitleInput from "./components/TitleInput/TitleInput";
import SaveTabButton from "./components/SaveTabButton/SaveTabButton";

const RiffinEditorDispatch = createContext(null);

const handleBlockSizeChange = (state, action)=> {
  const newInputs = utils.generateNewTextareaValueAfterSizeChange(action.type, action.OGBlock.inputs, " ", action.stepCount, action.OGBlock.cols);
  const newDashes = utils.generateNewTextareaValueAfterSizeChange(action.type, action.OGBlock.dashes, "-", action.stepCount, action.OGBlock.cols);
  // updates block proeprties
  const { cols, maxLength } = utils.generateNewSizePropertiesAfterSizeChange(action.type, action.OGBlock, action.stepCount);
  const indexOfBlockToUpdate = state.tablature.blocks.findIndex((block) => {
    if(action.OGBlock.tempKey && block.tempKey) {
      return (action.OGBlock.tempKey === block.tempKey);
    } else {
      return (action.OGBlock._id === block._id);
    }
  });
  state.tablature.blocks[indexOfBlockToUpdate] = {
    ...action.OGBlock,
    inputs: newInputs,
    dashes: newDashes,
    cols: cols,
    maxLength: maxLength,
  }
  return {
    tablature: state.tablature,
    selectedBlock: state.selectedBlock,
    cursor: state.cursor
  }
};

const handleAddCharacter = (state, action) => {
  const blockToAddTo = state.tablature.blocks[state.selectedBlock.index];
  blockToAddTo.inputs = utils.replaceTextareaValue(blockToAddTo.inputs, action.character, action.selectionStart)
  blockToAddTo.dashes = utils.replaceTextareaValue(blockToAddTo.dashes, " ", action.selectionStart)
  return {
    tablature: state.tablature,
    selectedBlock: state.selectedBlock,
    cursor: utils.generateCursorPositionObject(action.selectionStart + 1)
  }
};

const handleDeleteCharacter = (state, action) => {
  const blockToDeleteFrom = state.tablature.blocks[state.selectedBlock.index];
  blockToDeleteFrom.inputs = utils.replaceTextareaValue(blockToDeleteFrom.inputs, " ", action.selectionStart)
  blockToDeleteFrom.dashes = utils.replaceTextareaValue(blockToDeleteFrom.dashes, "-", action.selectionStart)
  return {
    tablature: state.tablature,
    selectedBlock: state.selectedBlock,
    cursor: utils.generateCursorPositionObject(action.selectionStart)
  }   
};

const handleDuplicateColumn = (state, action) => {
  const duplicationBlock = state.tablature.blocks[state.selectedBlock.index];
  for(let value of action.duplicationValues) {
    duplicationBlock.inputs = utils.replaceTextareaValue(duplicationBlock.inputs, value.inputValue, value.targetPosition)
    duplicationBlock.dashes = utils.replaceTextareaValue(duplicationBlock.dashes, value.dashValue, value.targetPosition)
  }
  return {
    tablature: state.tablature,
    selectedBlock: state.selectedBlock,
    cursor: utils.generateCursorPositionObject(action.selectionStart + 1)
  }
};

const handleUpdateSelection = (state, action) => {
  return {
    tablature: state.tablature,
    selectedBlock: utils.generateSelectedBlockObject(action.blockIndex, action.blockRef),
    cursor: utils.generateCursorPositionObject(action.selectionStart)
  };
};

const handleUpdateCursorPosition = (state, action) => {
  return {
    tablature: state.tablature,
    selectedBlock: state.selectedBlock,
    cursor: utils.generateCursorPositionObject(action.selectionStart)
  }
};

const handleDeleteColumn = (state, action) => {
  const deletionBlock = state.tablature.blocks[state.selectedBlock.index];
  for(let position of action.positionsToDelete) {
    deletionBlock.inputs = utils.replaceTextareaValue(deletionBlock.inputs, " ", position)
    deletionBlock.dashes = utils.replaceTextareaValue(deletionBlock.dashes, "-", position)
  }
  return {
    tablature: state.tablature,
    selectedBlock: state.selectedBlock,
    cursor: utils.generateCursorPositionObject(action.selectionStart - 1)
  }
};

const handleDeleteBlock = (state, action) => {
  state.tablature.blocks = state.tablature.blocks.filter((block) => {
    if(action.block.tempKey && block.tempKey) {
      return (action.block.tempKey !== block.tempKey) 
    } else {
      return (action.block._id !== block._id) 
    }
  });
  return {
    tablature: state.tablature,
    selectedBlock: state.selectedBlock,
    cursor: state.cursor
  }
};

const handleDuplicateBlock = (state, action) => {
  // * In React.StrictMode this causes 2 blocks to be pushed into the blocks array.
  const newBlock = { ...action.blockToDuplicate, tempKey: Date() + Math.random() };
  state.tablature.blocks.push(newBlock);
  return {
    tablature: state.tablature,
    selectedBlock: state.selectedBlock,
    cursor: state.cursor
  }
};

const handleAddNewBlock = (state, action) => {
  const newBlock = utils.getNewGuitarBlock(state.tablature.numberOfStrings)
  // * This check is a workaround to React.StrictMode's behavior of running reducers twice. Without this workaround, 2 blocks would be added.
  if(state.tablature.blocks.length < action.numberOfBlocksBeforeAdding + 1) {
    state.tablature.blocks.push(newBlock);
  }
  return {
    tablature: state.tablature,
    selectedBlock: state.selectedBlock,
    cursor: state.cursor
  }
};

const handleUpdateTablatureTitle = (state, action) => {
  state.tablature.name = action.name;
  return {
    tablature: state.tablature,
    selectedBlock: state.selectedBlock,
    cursor: state.cursor
  }
};

const handleUpdateBlockLabel = (state, action) => {
  state.tablature.blocks[action.index].label = action.value;
  return {
    tablature: state.tablature,
    selectedBlock: state.selectedBlock,
    cursor: state.cursor
  }
};

// TODO find a better way to access selected block rather than state.tablature.blocks[state.selectedBlock.index];
function riffinReducer(state, action) {
  switch (action.type) {
    case 'updateSelection':
      return handleUpdateSelection(state, action);
    case 'updateCursorPosition':
      return handleUpdateCursorPosition(state, action);
    case 'addCharacter':
      return handleAddCharacter(state, action);
    case 'deleteCharacter':
      return handleDeleteCharacter(state, action);
    case 'duplicateColumn':
      return handleDuplicateColumn(state, action);
    case 'deleteColumn':
      return handleDeleteColumn(state, action);
    case 'duplicateBlock':
      // * In React.StrictMode this causes 2 blocks to be pushed into the blocks array.
      return handleDuplicateBlock(state, action);
    case 'deleteBlock':
      return handleDeleteBlock(state, action);
    case 'increaseBlockSize':
      return handleBlockSizeChange(state, action)
    case 'decreaseBlockSize':
      return handleBlockSizeChange(state, action)
    case 'addNewBlock':
      return handleAddNewBlock(state, action);
    case 'updateTablatureTitle':
      return handleUpdateTablatureTitle(state, action);
    case 'updateBlockLabel':
      return handleUpdateBlockLabel(state, action);
    default:
      return {
        tablature: state.tablature,
        selectedBlock: state.selectedBlock,
        cursor: state.cursor
      }
  }
}

const RiffinEditor = (props) => {
  // Note: `dispatch` won't change between re-renders
  
  const [isLoading, setIsLoading] = useState(false);
  const [editor, dispatch] = useReducer(riffinReducer, {
    tablature: utils.getNewTablatureTemplateObject(props.numberOfStrings),
    selectedBlock: null,
    cursor: {position: null}
  });
  
  // Prevent crusor from jumping to end of input.
  useEffect(() => {
    if (editor.selectedBlock) {
      editor.selectedBlock.inputRef.current.selectionStart = editor.cursor.position;
      editor.selectedBlock.inputRef.current.selectionEnd = editor.cursor.position;
      console.log("New editor.cursor: ", editor.cursor.position);
    }
  }, [editor.cursor, editor.selectedBlock]);

  return (
    <RiffinEditorDispatch.Provider value={dispatch}>
      {!isLoading &&
      <>
        <TitleInput />
        <SaveTabButton tablature={editor.tablature} setIsLoading={setIsLoading} tags={props.tags}/>
        {editor.tablature.blocks.map((block, i) => (
          <TablatureBlock key={i} index={i} block={block} />)
        )}
        <AddTablatureBlockButton numberOfBlocks={editor.tablature.blocks.length} />
      </>
      }
    </RiffinEditorDispatch.Provider>
  );
}
 
export { RiffinEditor, RiffinEditorDispatch };