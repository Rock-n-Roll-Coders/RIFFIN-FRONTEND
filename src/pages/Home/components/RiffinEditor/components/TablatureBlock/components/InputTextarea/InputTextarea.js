// Components / hooks
import { useRef, useEffect, useState, useContext } from "react";
// Utils / services
import * as utils from "../../../../Utilities";

// MUI
import { useTheme } from "@mui/material/styles";
import { RiffinEditorDispatch } from "pages/Home/components/RiffinEditor/RiffinEditor";
import { DUPLICATION_COLUMN_GAP } from "pages/Home/components/RiffinEditor/EditorConfig";

const InputTextarea = (props) => {
  const [textAreaProperties] = useState({
    cols: props.block.cols,
    // TODO Refactor to numberOfStrings (stringCount is more tangled than youd think >.>)
    stringCount: props.block.numberOfStrings
  })
  const [mapOfLastColumnIndexes, setMapOfLastColumnIndexes] = useState(utils.getMapOfLastColumnIndexes(textAreaProperties))
  const [mapOfFirstColumnIndexes, setMapOfFirstColumnIndexes] = useState(utils.getMapOfFirstColumnIndexes(textAreaProperties))

  const dispatch = useContext(RiffinEditorDispatch);
  const ref = useRef();
  const theme = useTheme();
  
  const inputsStyle = {
    background: "transparent",
    margin: 0,
    fontFamily: "Fira Code",
    position: "relative",
    resize: "none",
    zIndex: 2,
    outline: "none",
    border: "none",
    color: theme.palette.primary.tabInput,
    fontSize: "1.2rem",
    padding: 0
  };

  // ------------- 🏗 MISC HELPERS --------------------------------
  
  /**
   * Checks if the cursor selection is on the last column.
   * @param {Number} selectionStart 
   * @param {Object} mapOfLastColumnIndexes 
   * @returns true if the cursor is on the last column, otherwise false
   */
  const selectionIsOnLastColumn = (selectionStart, mapOfLastColumnIndexes) => (selectionStart in mapOfLastColumnIndexes)

  /**
   * Checks if the cursor selection is on the First column.
   * @param {Number} selectionStart 
   * @param {Object} mapOfFirstColumnIndexes 
   * @returns true if the cursor is on the first column, otherwise false
   */
  const selectionIsOnFirstColumn = (selectionStart, mapOfFirstColumnIndexes) => (selectionStart in mapOfFirstColumnIndexes)

  /**
   * Checks if the cursor selection is on the second to last column. Second to last column is rarely needed and not stored in state.
   * @param {Number} selectionStart 
   * @param {Object} mapOfLastColumnIndexes 
   * @returns true if the cursor is on the second to last column, otherwise false
   */
  const selectionIsOnSecondToLastColumn = (selectionStart, mapOfLastColumnIndexes) => {
    const mapOfSecondToLastColumnIndexes = Object.keys(mapOfLastColumnIndexes).map((val) => val - 1)
    return mapOfSecondToLastColumnIndexes.includes(selectionStart)
  }

  // ------------- 🛠 HELPERS FOR DUPLICATING / DELETING ENTIRE COLUMNS --------------------------------

  /**
   * Gets the positions of the input textarea in a column directly above the curent selection.
   * @param {Number} selectionStart 
   * @param {Number} cols 
   * @returns an array of positions
   */
  const getPositionsAboveSelection = (selectionStart, cols) => {
    let arrayOfPositions = []
    const indexGapBetweenStrings = cols + 1 // The + 1 compensates for the hidden \n
    for(let position = selectionStart - indexGapBetweenStrings; position >= 0; position -= indexGapBetweenStrings) {
      arrayOfPositions.push(position - 1)
    }
    return arrayOfPositions
  }

  /**
   * Gets the positions of the input textarea in a column directly below the curent selection.
   * @param {Number} selectionStart 
   * @param {Number} cols 
   * @returns an array of positions
   */
  const getPositionsBelowSelection = (selectionStart, cols, editableLength) => {
    const indexGapBetweenStrings = cols + 1 // The + 1 compensates for the hidden \n
    let arrayOfPositions = []
    for(let position = selectionStart + indexGapBetweenStrings; position <= editableLength; position += indexGapBetweenStrings) {
      arrayOfPositions.push(position - 1)
    }
    return arrayOfPositions
  }
  
  /**
   * Gets the column of positions directly behind the cursor in the input textarea.
   * @param {Number} selectionStart 
   * @param {Number} cols 
   * @param {Number} maxLength 
   * @param {Number} numberOfStrings 
   * @returns 
   */
  const getColumnOfPositionsBeforeSelection = (selectionStart, cols, maxLength, numberOfStrings) => {
    let numberOfEditablePositionsOnBlock = maxLength - numberOfStrings
    let positionsAboveSelection = getPositionsAboveSelection(selectionStart - 1, cols)
    let positionsBelowSelection = getPositionsBelowSelection(selectionStart - 1, cols, numberOfEditablePositionsOnBlock)
    let allPositionsToDuplicate = [selectionStart - 2, ...positionsAboveSelection, ...positionsBelowSelection]
    return allPositionsToDuplicate
  }

  /**
   * Generates an array of objects represening the inputs to be duplicated and their target positions.
   * @param {Number} selectionStart 
   * @returns an array of objects
   */
  const generateDuplicationValues = (selectionStart) => {
    let positionsToDuplicate = getColumnOfPositionsBeforeSelection(selectionStart, textAreaProperties.cols, props.block.maxLength, textAreaProperties.stringCount)
    let inputsAsArray = [...props.block.inputs]
    const duplicationValues = [];
    positionsToDuplicate.forEach((position) => {
      const characterToDuplicate = inputsAsArray[position];
      duplicationValues.push({
        inputValue: characterToDuplicate,
        dashValue: (characterToDuplicate === " ") ? "-" : " ",
        targetPosition: position + DUPLICATION_COLUMN_GAP
      })
    })
    return duplicationValues;
  }
  // ------------- 🖐 HANDLERS --------------------------------

  /**
   * Sends a dispatch to update the selected block. The new selected block will be whichever block is handling the click.
   * @param {Object} event 
   */
  const handleClick = (event) => {
    const action = {
      type: "updateSelection",
      blockIndex: props.index,
      blockRef: ref,
      selectionStart: event.target.selectionStart
    };
    dispatch(action);
  };

  /**
   * Sends a dispatch to update the selection when an arrow key is pressed. This allows the user to move around the textarea with arrow keys.
   * @param {Object} event 
   */
  const handleKeyUp = (event) => {
    if(utils.isMovementKey(event.key)) {
      const action = {
        type: "updateCursorPosition",
        selectionStart: event.target.selectionStart
      };
      dispatch(action);
    }
  };

  /**
   * Sends a dispatch to insert a character into the textarea. This allows the user to type in the textarea on valid spaces. The only column that is not valid is the last column. This is because there are hidden new-line characters (\n) that keep the structure of the tab intact, making the textarea appear as if it were guitar strings.
   * @param {Number} selectionStart 
   * @param {String} key 
   */
  const handleAddCharacter = (selectionStart, key) => {
    let action = {};
    if(selectionIsOnLastColumn(selectionStart - 1, mapOfLastColumnIndexes)) {
      action = {
        type: 'updateCursorPosition',
        selectionStart: selectionStart - 1
      };
    } else {
     action ={ 
        type: "addCharacter",
        character: key,
        selectionStart: selectionStart - 1
      };
    }
    dispatch(action);
  }

  /**
   * Sends a dispatch to remove a character into the textarea. This allows the user to delete on valid spaces. The only column that is not valid is the first column. This prevents the user form deleting a new-line characters (\n) on the previous space, which keeps the structure of the tab intact.
   * @param {*} selectionStart 
   */
  const handleDeleteCharacter = (selectionStart) => {
    let action = {};
    if(selectionIsOnFirstColumn(selectionStart + 1, mapOfFirstColumnIndexes)) {
      action = {
        type: 'updateCursorPosition',
        selectionStart: selectionStart + 1
      };
    } else {
      action = {
        type: "deleteCharacter",
        selectionStart: selectionStart
      };
    }
    dispatch(action);
  }

  /**
   * Sends a dispatch to duplicate an entire column in the textarea. The values of the duplicated column will be pasted two spaces over, leaving one blank space between the original and duplicated columns.
   * @param {*} selectionStart 
   */
  const handleDuplicateColumn = (selectionStart) => {
    let action = {};
    if(selectionIsOnLastColumn(selectionStart - 1, mapOfLastColumnIndexes)
      || selectionIsOnFirstColumn(selectionStart - 1, mapOfFirstColumnIndexes)
      || selectionIsOnSecondToLastColumn(selectionStart - 1, mapOfLastColumnIndexes)
    ) {
      action = {
        type: 'updateCursorPosition',
        selectionStart: selectionStart - 1
      };
    } else {
      action = {
        type: "duplicateColumn",
        duplicationValues: generateDuplicationValues(selectionStart),
        selectionStart: selectionStart
      }
    }
    dispatch(action);
  }

  /**
   * Sends a dispatch to delete an entire column in the textarea.
   * @param {Number} selectionStart 
   */
  const handleDeleteColumn = (selectionStart) => {
    let action = {};
    if(selectionIsOnFirstColumn(selectionStart - 1, mapOfFirstColumnIndexes)) {
      action = {
        type: 'updateCursorPosition',
        selectionStart: selectionStart - 1
      };
    } else {
      const positionsToDelete = getColumnOfPositionsBeforeSelection(selectionStart, textAreaProperties.cols, props.block.maxLength, textAreaProperties.stringCount)
      action = {
        type: 'deleteColumn',
        positionsToDelete,
        selectionStart: selectionStart - 1
      }
    }
    dispatch(action);
  }

  /**
   * Checks the pressed key and routes to the appropriate dispatch handler.
   * @param {Object} event 
   */
  const handleChange = (event) => {
    event.preventDefault();
    const key = event.nativeEvent.data || "Backspace";
    const dispatchType = utils.getDispatchTypeOfPressedKey(key);
    if(dispatchType === undefined) {
      let action = {
        type: 'updateCursorPosition',
        selectionStart: event.target.selectionStart - 1
      };
      dispatch(action);
    } else if(dispatchType === "addCharacter") {
      handleAddCharacter(event.target.selectionStart, key)
    } else if(dispatchType === "deleteCharacter") {
      handleDeleteCharacter(event.target.selectionStart)
    } else if(dispatchType === "duplicateColumn") {
      handleDuplicateColumn(event.target.selectionStart)
    } else if(dispatchType === "deleteColumn") {
      handleDeleteColumn(event.target.selectionStart)
    }
  };

  // ------------- 🦾 EFFECTS --------------------------------
  
  /**
   * Updates the bounry maps when the size of the block changes.
   */
  useEffect(() => {
    setMapOfLastColumnIndexes(utils.getMapOfLastColumnIndexes(textAreaProperties))
    setMapOfFirstColumnIndexes(utils.getMapOfFirstColumnIndexes(textAreaProperties))
  }, [textAreaProperties, props.block.cols]);

  return (
    <textarea
      style={inputsStyle}
      value={props.block.inputs}
      onChange={handleChange}
      onKeyUp={handleKeyUp}
      onPaste={(event) => event.preventDefault()}
      onClick={handleClick}
      cols={textAreaProperties.cols}
      rows={textAreaProperties.stringCount}
      maxLength={props.block.maxLength}
      ref={ref}
    />
  );
}
 
export default InputTextarea;