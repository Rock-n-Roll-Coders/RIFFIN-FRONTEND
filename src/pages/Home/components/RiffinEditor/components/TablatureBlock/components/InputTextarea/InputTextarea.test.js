import { getNewGuitarBlock } from "pages/Home/components/RiffinEditor/Utilities";
import InputTextarea from "./InputTextarea";
import renderer from 'react-test-renderer';
// Note that style props may be undefined undefined because I'm not wrapping this in the theme context provider.

test('renders correctly using a new tablatureBlock', ()=> {
  const block = getNewGuitarBlock();
  const tree = renderer
    .create(<InputTextarea 
      block={block}
      index={0}
      numberOfStrings={6}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// Research how to mockup reducers.
test.todo('write unit tests for user interactions with the textarea')