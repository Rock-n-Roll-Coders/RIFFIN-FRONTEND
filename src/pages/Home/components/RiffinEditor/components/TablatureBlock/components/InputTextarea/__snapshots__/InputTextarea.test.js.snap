// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly using a new tablatureBlock 1`] = `
<textarea
  cols={40}
  maxLength={251}
  onChange={[Function]}
  onClick={[Function]}
  onKeyUp={[Function]}
  onPaste={[Function]}
  rows={6}
  style={
    Object {
      "background": "transparent",
      "border": "none",
      "color": undefined,
      "fontFamily": "Fira Code",
      "fontSize": "inherit",
      "margin": 0,
      "outline": "none",
      "padding": 0,
      "position": "relative",
      "resize": "none",
      "zIndex": 2,
    }
  }
  value="|                                      |
|                                      |
|                                      |
|                                      |
|                                      |
|                                      |"
/>
`;
